vid/src/avid_m/namespaced.rs

```diff
@@ -0,0 +1,241 @@
+//! This file implements the namespaced AvidM scheme.
+
+use super::{AvidMCommit, AvidMShare, RawAvidMShare};
+use crate::{
+    avid_m::{AvidMScheme, MerkleTree},
+    VidError, VidResult, VidScheme,
+};
+use jf_merkle_tree::MerkleTreeScheme;
+use serde::{Deserialize, Serialize};
+use std::ops::Range;
+
+/// Dummy struct for namespaced AvidM scheme
+pub struct NsAvidMScheme;
+
+/// Namespaced commitment type
+pub type NsAvidMCommit = super::AvidMCommit;
+/// Namespaced parameter type
+pub type NsAvidMParam = super::AvidMParam;
+
+/// Namespaced share for each storage node
+#[derive(Clone, Debug, Hash, Serialize, Deserialize, Eq, PartialEq, Default)]
+pub struct NsAvidMShare {
+    /// Index number of the given share.
+    index: u32,
+    /// The list of all namespace commitments
+    ns_commits: Vec<AvidMCommit>,
+    /// The size of each namespace
+    ns_lens: Vec<usize>,
+    /// Actual share content
+    content: Vec<RawAvidMShare>,
+}
+
+impl NsAvidMShare {
+    fn ns_share(&self, ns_id: usize) -> AvidMShare {
+        AvidMShare {
+            index: self.index,
+            payload_byte_len: self.ns_lens[ns_id],
+            content: self.content[ns_id].clone(),
+        }
+    }
+
+    /// Return the length of underlying payload in bytes
+    pub fn payload_byte_len(&self) -> usize {
+        self.ns_lens.iter().sum()
+    }
+}
+
+impl NsAvidMScheme {
+    /// Setup an instance for AVID-M scheme
+    pub fn setup(recovery_threshold: usize, total_weights: usize) -> VidResult<NsAvidMParam> {
+        NsAvidMParam::new(recovery_threshold, total_weights)
+    }
+
+    /// Commit to a payload given namespace table.
+    pub fn commit(
+        param: &NsAvidMParam,
+        payload: &[u8],
+        ns_table: impl IntoIterator<Item = Range<usize>>,
+    ) -> VidResult<NsAvidMCommit> {
+        let ns_commits = ns_table
+            .into_iter()
```

Comment:
should we validate `ns_table` before committing them? better yet, can we rely on type safety, i remember `NsTable` struct can be constructed safely from an iter of ranges with validation, and we can just accept a `NsTable` type here.

Reply:
Again `NsTable` currently resides in `sequencer` scope.
Let's handle it together in #2458